lib/engine-universal/src/code_memory.rs

```diff
@@ -2,217 +2,235 @@
+    pub fn writable_address(&self, offset: usize) -> *const u8 {
         unsafe {
-            region::protect(
-                self.mmap.as_mut_ptr(),
-                self.start_of_nonexecutable_pages,
-                region::Protection::READ_EXECUTE,
-            )
+            // TODO: encapsulate offsets so that this `offset` is guaranteed to be sound.
+            self.map.offset(offset as isize)
         }
-        .expect("unable to make memory readonly and executable");
     }
+}
 
-    /// Calculates the allocation size of the given compiled function.
-    fn function_allocation_size(func: FunctionBodyRef<'_>) -> usize {
-        match &func.unwind_info {
-            Some(CompiledFunctionUnwindInfoRef::WindowsX64(info)) => {
-                // Windows unwind information is required to be emitted into code memory
-                // This is because it must be a positive relative offset from the start of the memory
-                // Account for necessary unwind information alignment padding (32-bit alignment)
-                ((func.body.len() + 3) & !3) + info.len()
+impl Drop for CodeMemory {
+    fn drop(&mut self) {
+        if let Some(source_pool) = self.source_pool.take() {
+            drop(source_pool.push(CodeMemory {
+                source_pool: None,
+                map: self.map,
+                size: self.size,
+                executable_size: 0,
+            }));
+        } else {
+            unsafe {
+                // TODO(nagisa): log errors here.
+                drop(mm::munmap(self.map.cast(), self.size));
             }
-            _ => func.body.len(),
         }
     }
+}
 
-    /// Copies the data of the compiled function to the given buffer.
-    ///
-    /// This will also add the function to the current function table.
-    fn copy_function<'a>(
-        registry: &mut UnwindRegistry,
-        func: FunctionBodyRef<'_>,
-        buf: &'a mut [u8],
-    ) -> &'a mut [VMFunctionBody] {
-        assert_eq!(buf.as_ptr() as usize % ARCH_FUNCTION_ALIGNMENT, 0);
-
-        let func_len = func.body.len();
-
-        let (body, remainder) = buf.split_at_mut(func_len);
-        body.copy_from_slice(&func.body);
-        let vmfunc = Self::view_as_mut_vmfunc_slice(body);
-
-        if let Some(CompiledFunctionUnwindInfoRef::WindowsX64(info)) = &func.unwind_info {
-            // Windows unwind information is written following the function body
-            // Keep unwind information 32-bit aligned (round up to the nearest 4 byte boundary)
-            let unwind_start = (func_len + 3) & !3;
-            let unwind_size = info.len();
-            let padding = unwind_start - func_len;
-            assert_eq!((func_len + padding) % 4, 0);
-            let slice = remainder.split_at_mut(padding + unwind_size).0;
-            slice[padding..].copy_from_slice(&info);
-        }
+unsafe impl Send for CodeMemory {}
+unsafe impl Sync for CodeMemory {}
```

Comment:
So the most plausible reason `CodeMemory` should be `Sync` is the

```
impl<T> Send for Arc<T>
    where T: Sync + Send + ?Sized {}
```

implementation. Unless `CodeMemory` is `Sync` (which it is, being pretty largely isomorphic with `Box<[u8]>`), many of the types storing `CodeMemory` indirectly won’t be `Send`. In particular `wasmer::Module`/`wasmer::Instance` have a `Send` assertion too. We could look into making these not-Send, but that seems like it would be out of scope for this PR, especially given that the former `CodeMemory` was also `Sync`.

Reply:
Great! Also thinking about it some more I guess my biggest fear about CodeMemory being Sync would not be related to it being Sync but to the returned raw pointers being used in the wrong thread, which could still happen given we have a Send bound there that we can't get rid of as we want to be able to reuse CodeMemory.

TLDR: I'd feel great if it were possible to have one fewer instance of unsafe in our code, but if there's any issue in landing that in nearcore feel free to re-add the unsafe impl as it's probably not a big deal!